Configure NLogProviderOptions from appsettings.json

snakefoot opened the pull request:
Added support for configuring NLogProviderOptions from appsettings.json (Updated package "Microsoft.Extensions.Logging" from Version="2.0.1" to Version="2.1.0")

Added new NLogLoggerProvider-constructor with LogFactory-parameter

304NotModified commented:
I was just preparing 1.3 ;)

https://github.com/NLog/NLog.Extensions.Logging/pull/247

304NotModified commented:
I don't fully understand what this PR is changing. What's the benefit here?

304NotModified commented on src/NLog.Extensions.Configuration/ConfigSettingLayoutRenderer.cs:
```diff
@@ -126,33 +125,48 @@ private IConfigurationRoot LoadFileConfiguration(string fileNames)
+        private static bool IsDefaultConfiguration(Layout fileName, string defaultFileName)
+        {
+            var simpleLayout = fileName as SimpleLayout;
+            return simpleLayout == null || string.IsNullOrEmpty(simpleLayout.Text) || ReferenceEquals(simpleLayout.Text, defaultFileName);
```
I still prefer pattern matching, but it's no show-stopper

304NotModified commented on src/NLog.Extensions.Configuration/NLog.Extensions.Configuration.csproj:
```diff
@@ -12,7 +12,10 @@
-    <Description>NLog extensions for Microsoft.Extensions.Configuration</Description>
+    <Description>NLog extensions for Microsoft.Extensions.Configuration
+
+Read appsettings.json on Net Core using ${configsetting:name=MySetting}
```
I prefer writing .NET instead of Net and keep it consistent everywhere. 

I think it's everywhere the platforms are written as:

- .NET
- .NET Standard
- .NET Core
- ASP.NET
- ASP.NET Core

snakefoot replied:
Done

snakefoot commented:
> I don't fully understand what this PR is changing. What's the benefit here?

Not a big AspNet-Core-Expert, but I see lots of people doing this:

```
   var builder = new ConfigurationBuilder()
        .SetBasePath(env.ContentRootPath)
        .AddJsonFile("appsettings.json", optional: false, reloadOnChange: true)
        .AddJsonFile($"appsettings.{env.EnvironmentName}.json", optional: true)
        .AddEnvironmentVariables();
```

Trying to make this layoutrenderer do the same when default configuration.

snakefoot commented:
> shouldn't we do .SetBasePath(env.ContentRootPath) then always?

Where do I get the hosting environment?

304NotModified commented:
> Where do I get the hosting environment?

Good one.

And always  AppDomain.BaseDirectory then? I think it's a bit magic this only done with the "default config". Or is that a breaking (behavior) change?

snakefoot commented:
Btw. going out now, so will not be able to make any changes. If you need more modifications, then you to push them yourself

> Or is that a breaking (behavior) change?

Breaking change from what?

304NotModified commented:
OK have fun

304NotModified commented:
I can't push on your master :|

snakefoot commented:
@304NotModified What changes do you like me to make?

304NotModified commented on src/NLog.Extensions.Configuration/ConfigSettingLayoutRenderer.cs:
```diff
@@ -126,33 +125,48 @@ private IConfigurationRoot LoadFileConfiguration(string fileNames)
             }
 
+            if (includeEnvironmentVariables)
+                configBuilder.AddEnvironmentVariables();
```
proposal: 

- create a setting for this 
- remove the IsDefaultConfiguration
- disabled by default?

snakefoot replied:
Ok I will remove it then

snakefoot commented:
@304NotModified Now changed NLog.Extension.Configuration to have single dependency. And added extension methods for making it easy to register configuration.

Also changed NLog.Extension.Hosting to depend on NLog.Extension.Configuration.

snakefoot commented:
Btw. noticed that [Microsoft.Extensions.Logging ver. 2.1](https://www.nuget.org/packages/Microsoft.Extensions.Logging/2.1.0) introduces dependency on `Microsoft.Extensions.Configuration.Abstractions`.

If we ever upgrade from ver. 2.0 to 2.1, then NLog.Extensions.Configuration can merge into Microsoft.Extensions.Logging for NetStandard2.0 and Net461

304NotModified commented:
> If we ever upgrade from ver. 2.0 to 2.1,

Upgrade is fine to me. I think supporting the last RTM is reasonable if there are changes we need'

snakefoot commented:
> Upgrade is fine to me. I think supporting the last RTM is reasonable if there are changes we need'

Looks like this new feature https://github.com/aspnet/Logging/pull/723 (NetCore2.1) is making it hard for NLog to be a standard ILoggerProvider that lives side-by-side others.

If NLogLoggerProvider does not support ISupportExternalScope, then the default LoggerFactory will introduce a huge penalty and allocate like crazy. See also  #714

snakefoot commented:
@304NotModified Have now merged NLog.Extension.Configuration together with NLog.Extension.Logging. Also reported some issues discovered in the upgrade to NetCore2.1

snakefoot commented:
@304NotModified Hard for me to decide if NLog.Extension.Logging should be upgraded to depend on Microsoft.Extensions.Logging ver. 2.1

Or if it should stay at Microsoft.Extensions.Logging ver. 2.0 or keep NLog.Extension.Configuration-project. Seldom like when component-libraries starts to push requirements.

snakefoot commented:
> There is always the choice not to update NLog

Yes it will probably also push for NLog 4.6 that provides some new stuff like:

https://github.com/NLog/NLog/pull/2949 - Reduce Allocations
https://github.com/NLog/NLog/pull/2910 - Remove NDLC-ToString-Allocation on NetFramework
https://github.com/NLog/NLog/pull/2891 - Add support for NLog.config in AppSettings.json

304NotModified commented:
PS: I always try to update the NLog dependency in the package :)

304NotModified commented:
Is this a breaking change?

(sorry haven't found time to review the larger PRs)

snakefoot commented:
> Is this a breaking change?

I guess it is a breaking change of NLog.Extension.Configuration as it is now gone.

304NotModified commented on examples/NetCore2/ConsoleExample/Program.cs:
```diff
@@ -44,21 +44,20 @@ private static IServiceProvider BuildDi()
-                CaptureMessageProperties = true
+                loggingBuilder.ClearProviders();
+                loggingBuilder.SetMinimumLevel(Microsoft.Extensions.Logging.LogLevel.Trace);
+                loggingBuilder.AddNLog(config);
```
I'm not sure if this is DI-proof. 

I think it's a good change to move to the Factory instead of Provider, that's the one we could also bootstrap from DI. See https://github.com/NLog/NLog.Extensions.Logging/pull/253

snakefoot replied:
Please explain DI-proof?

304NotModified replied:
I don't like to change the call, if we need a new dependency. But is a bit debatable if the config is here an dependency or not.

snakefoot replied:
I was planning that NLogLoggingProvider.Options and NLog-LoggingConfiguration was loaded from IConfiguration. So it would be natural to receive the config when doing AddNLog.

But not that skilled with DI and fluent-interface, so please help with creating a great interface.

greenmooseSE commented on src/NLog.Extensions.Logging/Logging/NLogLoggerProvider.cs:
```diff
@@ -34,7 +40,18 @@ public NLogLoggerProvider()
+        /// </summary>
+        /// <param name="options"></param>
+        /// <param name="logFactory">Optional isolated NLog LogFactory</param>
+        public NLogLoggerProvider(NLogProviderOptions options, LogFactory logFactory)
```
Shouldn't this constructor be reachable via ConfigureExtensions ? (this constructor seems to never be called explicitly, only implicitly via ```public NLogLoggerProvider(NLogProviderOptions options) : this(options, null)```).

snakefoot replied:
I was thinking that people who needed to provide an isolated LogFactory was in a special unit-testing scenario, and didn't need any help from ConfigureExtensions

304NotModified commented:
I'm sorry this PR is still open. It's a bit too big for me to handle currently.

snakefoot commented:
Now just provides the ability to provide isolated LogFactory-object.

304NotModified commented on src/NLog.Extensions.Hosting/Extensions/ConfigureExtensions.cs:
```diff
@@ -41,7 +41,17 @@ public static IHostBuilder UseNLog(this IHostBuilder builder, NLogProviderOption
+                    var provider = new NLogLoggerProvider(options ?? new NLogProviderOptions());
+                    if (hostbuilder.Configuration != null)
+                    {
+                        // TODO ConfigSettingLayoutRenderer.DefaultConfiguration = hostbuilder.Configuration;
```
when should be do this TODO?

snakefoot replied:
Next PR. You just requested me to split this into more PRs

304NotModified commented on src/NLog.Extensions.Logging/Extensions/ConfigureExtensions.cs:
```diff
@@ -33,6 +35,19 @@ public static ILoggerFactory AddNLog(this ILoggerFactory factory, NLogProviderOp
         }
 
+        /// <summary>
+        /// Enable NLog as logging provider in .NET Core.
```
not sure if this should be ".NET Standard." instead of ".NET Core"

snakefoot replied:
NetCore is the platform. NetStandard is the API-specification.

snakefoot replied:
But yes lets call it Net Standard since this packages works on both NetCore and NetFramework.

snakefoot replied:
Have updated all comments to say "for Microsoft Extension Logging".

304NotModified commented on src/NLog.Extensions.Logging/Extensions/ConfigureExtensions.cs:
```diff
@@ -83,5 +111,48 @@ public static LoggingConfiguration ConfigureNLog(this ILoggerFactory loggerFacto
+            var provider = new NLogLoggerProvider(new NLogProviderOptions());
+            if (configuration != null)
+            {
+                // TODO ConfigSettingLayoutRenderer.DefaultConfiguration = configuration;
```
is this TODO for this PR, or do we need a Github issue?

snakefoot replied:
Next PR. You just requested me to split this into more PRs

304NotModified commented on test/NLog.Extensions.Logging.Tests/CustomBeginScopeTest.cs:
```diff
@@ -40,7 +40,7 @@ public void TestNonSerializableSayHi()
             Assert.Single(target.Logs);
             Assert.Equal("Hi Earth. Welcome Earth People", target.Logs[0]);
-            Assert.Equal("Earth People", scopeString);
+            // Assert.Equal("Earth People", scopeString); <-- Bug https://github.com/aspnet/Logging/issues/893
```
Dunno if this is solved yet, as the repos has been changed.

snakefoot replied:
Think it has been resolved in NetCore2.2, but that is not the long-term-support-version (NetCore 2.1 is)

snakefoot commented on src/NLog.Extensions.Logging/Extensions/ConfigureExtensions.cs:
```diff
@@ -83,5 +153,48 @@ public static LoggingConfiguration ConfigureNLog(this ILoggerFactory loggerFacto
         }
+
+        /// <summary>
+        /// Factory method for <see cref="NLogLoggerProvider"/>
```
Nope it is an extension-method with fluent-interface.

snakefoot commented on src/NLog.Extensions.Logging/Extensions/ConfigureExtensions.cs:
```diff
@@ -83,5 +153,48 @@ public static LoggingConfiguration ConfigureNLog(this ILoggerFactory loggerFacto
+        /// <param name="nlogProvider"></param>
+        /// <param name="configuration">Microsoft Extension Configuration</param>
+        /// <returns></returns>
+        public static NLogLoggerProvider ConfigureNLogProvider(this NLogLoggerProvider nlogProvider, IConfiguration configuration)
```
Accepted
